theano/gpuarray/dnn.py

```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    __props__ = ('precision',)
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+
+    def __init__(self, precision):
+        DnnBase.__init__(self, ["c_code/spatialtf_grid.c"], "spatialtf_grid")
+
+        self.precision = precision
+
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
+
+    def make_node(self, grid_dimensions, theta, desc):
+        context_name = infer_context_name(desc, theta)
+
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
+        assert theta.ndim == 3
+
+        # Allocate GPU memory for grid of coordinates
+        grid = GpuArrayType(dtype=self.precision,
+                            broadcastable=(False, False, False, False,),
+                            context_name=context_name)()
+
+        return Apply(self, [grid_dimensions, theta, desc], [grid])
+
+    def L_op(self, inputs, outputs, output_grads):
+        pass
+
+
+class GpuDnnGridSampler(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid sampler for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('precision',)
+    _cop_num_inputs = 6
+    _cop_num_outputs = 1
+
+    def __init__(self, precision):
+        DnnBase.__init__(self, ["c_code/spatialtf_sampler.c"], "spatialtf_sampler")
+
+        self.precision = precision
+
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
+
+    def make_node(self, img, grid, grid_dimensions, desc,
+                  alpha=None, beta=None):
+        context_name = infer_context_name(img, grid)
+
+        img = as_gpuarray_variable(img, context_name)
+        grid = as_gpuarray_variable(grid, context_name)
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+
+        output = GpuArrayType(dtype=self.precision,
+                              broadcastable=img.type.ndim * (False,),
+                              context_name=context_name)()
+
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+        if output.type.ndim != 4:
+            raise TypeError('output must be a 4D tensor')
+
+        if img.type.ndim != output.type.ndim:
+            raise TypeError('The number of dimensions of img and output must match')
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        alpha = ensure_dt(alpha, _one, 'alpha', img.dtype)
+        beta = ensure_dt(beta, _zero, 'beta', img.dtype)
+
+        return Apply(self, [img, grid, grid_dimensions, desc, alpha, beta],
+                     [output])
+
+    def L_op(self, inputs, outputs, output_grads):
+        pass
+
+
+def dnn_spatialtf(img, theta, grid_dims, alpha=None, beta=None, precision=None):
```

Comment:
`precision` seems to be in fact the dtype of output. And I was thinking that the dtype of output should just be the dtype of input. What do you think ?

In any case, rename it here to dtype, output_dtype (maybe the best name ?) or output_type.